jolt-core/src/zkvm/prover.rs

```diff
@@ -0,0 +1,1420 @@
+        );
+        let mut ram_output_check = OutputSumcheckProver::gen(
+            &self.initial_ram_state,
+            &self.final_ram_state,
+            &self.program_io,
+            self.ram_K,
+            &mut self.transcript,
+        );
+
+        #[cfg(feature = "allocative")]
+        {
+            let mut flamegraph = FlameGraphBuilder::default();
+            for sumcheck in stage2_instances.iter() {
+                sumcheck.update_flamegraph(&mut flamegraph);
+            }
+            write_flamegraph_svg(flamegraph, "stage2_start_flamechart.svg");
+        }
+
+        tracing::info!("Stage 2 proving");
+        let (sumcheck_proof, _r_stage2) = BatchedSumcheck::prove(
+            vec![
+                &mut spartan_inner,
+                &mut spartan_product_virtual_remainder,
+                &mut ram_raf_evaluation,
+                &mut ram_read_write_checking,
+                &mut ram_output_check,
+            ],
+            &mut self.opening_accumulator,
+            &mut self.transcript,
+        );
+
+        #[cfg(feature = "allocative")]
+        {
+            let mut flamegraph = FlameGraphBuilder::default();
+            for sumcheck in stage2_instances.iter() {
+                sumcheck.update_flamegraph(&mut flamegraph);
+            }
+            write_flamegraph_svg(flamegraph, "stage2_end_flamechart.svg");
+        }
+
+        // Drop product immediately because it has lifetimes.
+        drop(spartan_product_virtual_remainder);
+        drop_in_background_thread((
+            spartan_inner,
+            ram_raf_evaluation,
+            ram_read_write_checking,
+            ram_output_check,
+        ));
+
+        (first_round_proof, sumcheck_proof)
+    }
+
+    fn prove_stage3(&mut self) -> SumcheckInstanceProof<F, ProofTranscript> {
+        #[cfg(not(target_arch = "wasm32"))]
+        print_current_memory_usage("Stage 3 baseline");
+        let span = tracing::span!(tracing::Level::INFO, "Stage 3 sumchecks");
+        let _guard = span.enter();
+
+        let mut spartan_shift = ShiftSumcheckProver::gen(
+            &self.trace,
+            &self.preprocessing.bytecode,
+            &self.opening_accumulator,
+            &mut self.transcript,
+        );
+        let mut spartan_instruction_input = InstructionInputSumcheckProver::gen(
+            &self.trace,
+            &self.opening_accumulator,
+            &mut self.transcript,
+        );
+        let mut spartan_product_virtual_claim_check =
+            ProductVirtualInnerProver::new(&self.opening_accumulator, &mut self.transcript);
+        let instruction_ra_evals =
+            instruction_lookups::compute_ra_evals(&self.trace, &self.opening_accumulator);
+        let mut lookups_ra_hamming_weight = instruction_lookups::gen_ra_hamming_weight_prover(
+            &instruction_ra_evals,
+            &mut self.transcript,
+        );
+        self.instruction_ra_evals = Some(instruction_ra_evals);
```

Comment:
perhaps we should just move this hamming weight sumcheck to Stage 6 (alongside the instruction booleanity sumcheck) so that we don't have to cache `instruction_ra_evals` like this. The memory footprint of this Hamming weight sumcheck is negligible so we should be able to move it around freely